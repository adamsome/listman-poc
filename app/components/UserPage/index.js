import React from 'react'

import Tiles from '../Tiles'
import UserProfile from '../UserProfile'
import BodyColumn from '../Layout/BodyColumn'
import Container from '../Layout/Container'

import './UserPage.scss'

const UserPage = ({ user, lists, isLoading }) => (
  <BodyColumn>
    <UserProfile user={user} isLoading={isLoading} />
    <Container spaceAbove>
      <h2 className={"title is-2" + ((isLoading)
        ? " user-page__lists-heading--loading"
        : ""
      )}>
        Lists
      </h2>
      <Tiles entities={lists} isLoading={isLoading} />
    </Container>
  </BodyColumn>
)

export default UserPage
