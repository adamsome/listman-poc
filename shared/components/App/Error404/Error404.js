import React from 'react';

import BodyColumn from '../../Layout/BodyColumn'

const Error404 = ({ error }) => (
  <BodyColumn>
    <h1 className="title is-1">
      Page not found!
    </h1>
    <p className="subtitle is-5">
      {error}
    </p>
  </BodyColumn>
)

export default Error404;
