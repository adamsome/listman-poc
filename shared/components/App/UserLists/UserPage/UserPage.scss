.user-page__lists-heading--loading {
  color: #e6e6e6;
}
