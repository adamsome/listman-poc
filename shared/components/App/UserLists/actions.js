import { normalize } from 'normalizr'

import * as fromApp from '../../../reducers'
import * as fromUsers from '../../../reducers/users'
import * as schema from './schemas'

const dispatchError = ({ dispatch, err, ...rest }) => {
  const unknown = !err.response || !err.response.data.error
  const error = (unknown) ? err : err.response.data.error
  const payload = error.message + ((unknown) ? ' (Unknown error)' : '')
  console.error(payload, error)
  dispatch({ ...rest, status: 'error', payload })
  throw err
}

// Fetch User

export const fetchUser = (username) => (dispatch, _, { axios }) => {
  const type = 'FETCH_USER'
  dispatch({ type, username })

  return axios.get(`/users/${username}`)
    .then(({ data }) => {
      const payload = normalize(data, schema.user)
      dispatch({ type, username, status: 'success', payload })
    })
    .catch((err) => dispatchError({ dispatch, err, type, username }))
}

const shouldFetchUser = (state, username) => {
  const user = fromUsers.getUser(fromApp.getUsers(state), username)
  if (!user || fromApp.getListsByUserError(state, username)) {
    return true
  } else if (fromApp.getListsByUserIsLoading(state, username)) {
    return false
  }
  // TODO: Support user invalidation (i.e. `return user.didInvalidate`)
  return false
}

export const fetchUserIfNeeded = (username) => (dispatch, getState) => (
  (shouldFetchUser(getState(), username))
    ? dispatch(fetchUser(username))
    : Promise.resolve()
)

// Fetch User Lists

export const fetchUserLists = (username) => (dispatch, _, { axios }) => {
  const type = 'FETCH_USER_LISTS'
  dispatch({ type, username })

  return axios.get(`/users/${username}/lists`)
    .then(({ data }) => {
      const payload = normalize(data, schema.listArray)
      dispatch({ type, username, status: 'success', payload })
    })
    .catch((err) => dispatchError({ dispatch, err, type, username }))
}

const shouldFetchUserLists = (state, username) => {
  const userLists = fromApp.getListsByUser(state, username)
  if (!userLists || fromApp.getListsByUserError(state, username)) {
    return true
  } else if (fromApp.getListsByUserIsLoading(state, username)) {
    return false
  }
  // TODO: Support userList invalidation (i.e. `return userList.didInvalidate`)
  return false
}

export const fetchUserListsIfNeeded = (username) => (dispatch, getState) => (
  (shouldFetchUserLists(getState(), username))
    ? dispatch(fetchUserLists(username))
    : Promise.resolve()
)

// Add List

export const addList = (username, name) => (dispatch, _, { axios }) => {
  const type = 'ADD_LIST'
  dispatch({ type, username })

  return axios.post(`/users/${username}/lists`, { name })
    .then(({ data }) => {
      const payload = normalize(data, schema.list)
      dispatch({ type, username, status: 'success', payload })
    })
    .catch((err) => dispatchError({ dispatch, err, type, username }))
}
