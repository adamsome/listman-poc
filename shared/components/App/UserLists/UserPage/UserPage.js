import React from 'react'

import TileList from '../../../TileList'
import UserProfile from '../../../UserProfile'
import BodyColumn from '../../../Layout/BodyColumn'
import Container from '../../../Layout/Container'

import './UserPage.scss'

const UserPage = ({ user, lists, isLoading,
                    isAdding, addError, addList }) => (
  <BodyColumn>
    <UserProfile user={user} isLoading={isLoading} />
    <Container spaceAbove>
      <h2 className={"title is-2 " + ((isLoading)
        ? " user-page__lists-heading--loading"
        : ""
      )}>
        Lists
      </h2>
      {
        (addError)
          ? <article className="message is-danger">
              <div className="message-body">
                <strong>Error: </strong>{addError}
              </div>
            </article>
          : null
      }
      <TileList
        entities={lists}
        isLoading={isLoading}
        isAdding={isAdding}
        addError={addError}
        add={addList}
      />
    </Container>
  </BodyColumn>
)

export default UserPage
