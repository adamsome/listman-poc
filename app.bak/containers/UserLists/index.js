import React from 'react'
import { connect } from 'react-redux';

import * as fromApp from '../../reducers'
import * as fromUsers from '../../reducers/users'
import { fetchUserIfNeeded, fetchUserListsIfNeeded, addList } from './actions'
import UserPage from '../../components/UserPage'
import NotFound from '../../components/NotFound'

export class UserLists extends React.Component {
  componentDidMount() {
    const { username } = this.props
    this.fetchUserLists(username)
  }

  componentDidUpdate(prevProps) {
    // Fetch if user changed (or if previous user was undefined)
    const { username } = this.props
    const prevUsername = prevProps.username
    if (username !== prevUsername) {
      console.log(`<UserLists> update fetch` +
                  `(user ${username} != prevUser ${prevUsername})`)
      this.fetchUserLists(username)
    }
  }

  fetchUserLists(username) {
    this.props.fetchUserListsIfNeeded(username)
      .then(() => this.props.fetchUserIfNeeded(username))
      .catch(e => e) // Actions handle the errors
  }

  render() {
    const { username, user, lists, isLoading,
            error, isAdding, addError, addList } = this.props
    return (error)
      ? <NotFound error={error} />
      : <UserPage
          user={user}
          lists={lists}
          isLoading={isLoading}
          isAdding={isAdding}
          addError={addError}
          addList={(name) => addList(username, name)}
        />
  }
}

const mapStateToProps = (state, ownProps) => {
  const username = ownProps.params.username
  return {
    username,
    user: fromUsers.getUser(fromApp.getUsers(state), username),
    lists: fromApp.getListsByUser(state, username),
    isLoading: fromApp.getListsByUserIsLoading(state, username),
    error: fromApp.getListsByUserError(state, username),
    isAdding: fromApp.getListsByUserIsAdding(state, username),
    addError: fromApp.getListsByUserAddError(state, username),
  }
}

export default connect(
  mapStateToProps,
  { fetchUserIfNeeded, fetchUserListsIfNeeded, addList },
)(UserLists)

