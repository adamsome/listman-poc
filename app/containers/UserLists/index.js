import React from 'react'
import { connect } from 'react-redux';

import * as fromApp from '../../reducers'
import * as fromUsers from '../../reducers/users'
import { fetchUserListsIfNeeded, addList } from './actions'
import UserPage from '../../components/UserPage'
import NotFound from '../../components/NotFound'

export class UserLists extends React.Component {
  componentDidMount() {
    const { userID } = this.props
    this.fetchUserLists(userID)
  }

  componentDidUpdate(prevProps) {
    // Fetch if user changed (or if previous user was undefined)
    const { userID } = this.props
    const prevUserID = prevProps.userID
    if (userID !== prevUserID) {
      console.log(`<UserLists> update fetch` +
                  `(user ${userID} != prevUser ${prevUserID})`)
      this.fetchUserLists(userID)
    }
  }

  fetchUserLists(userID) {
    this.props.fetchUserListsIfNeeded(userID)
  }

  render() {
    const { user, lists, isLoading, error } = this.props
    return (error)
      ? <NotFound error={error} />
      : <UserPage user={user} lists={lists} isLoading={isLoading} />
  }
}

const mapStateToProps = (state, ownProps) => {
  const userID = fromApp.getCurrentUserID(state, ownProps)
  return {
    userID,
    user: fromUsers.getUser(fromApp.getUsers(state), userID),
    lists: fromApp.getListsByUser(state, userID),
    isLoading: fromApp.getListsByUserIsLoading(state, userID),
    error: fromApp.getListsByUserError(state, userID),
  }
}

export default connect(
  mapStateToProps,
  { fetchUserListsIfNeeded, addList },
)(UserLists)

