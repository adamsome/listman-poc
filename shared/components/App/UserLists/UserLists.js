import React from 'react'
import { connect } from 'react-redux';

import * as fromApp from '../../../reducers'
import * as fromUsers from '../../../reducers/users'
import { fetchUserIfNeeded, fetchUserListsIfNeeded, addList } from './actions'
import UserPage from './UserPage'
import Error404 from '../../Error404'

export class UserLists extends React.Component {
  componentDidMount() {
    const { username } = this.props
    this.fetchUserLists(username)
  }

  componentDidUpdate(prevProps) {
    // Fetch if user changed (or if previous user was undefined)
    const { username } = this.props
    const prevUsername = prevProps.username
    if (username !== prevUsername) {
      console.log(`<UserLists> update fetch` +
                  `(user ${username} != prevUser ${prevUsername})`)
      this.fetchUserLists(username)
    }
  }

  fetchUserLists(username) {
    this.props.fetchUserListsIfNeeded(username)
      .then(() => this.props.fetchUserIfNeeded(username))
      .catch(e => e) // Actions handle the errors
  }

  render() {
    const { username, user, lists, isLoading,
            error, isAdding, addError, addList } = this.props
    return (error)
      ? <Error404 error={error} />
      : <UserPage
          user={user}
          lists={lists}
          isLoading={isLoading}
          isAdding={isAdding}
          addError={addError}
          addList={(name) => addList(username, name)}
        />
  }
}

const mapStateToProps = (state, { params: { username } }) => ({
  username,
  user: fromUsers.getUser(fromApp.getUsers(state), username),
  lists: fromApp.getListsByUser(state, username),
  isLoading: fromApp.getListsByUserIsLoading(state, username),
  error: fromApp.getListsByUserError(state, username),
  isAdding: fromApp.getListsByUserIsAdding(state, username),
  addError: fromApp.getListsByUserAddError(state, username),
})

export default connect(
  mapStateToProps,
  { fetchUserIfNeeded, fetchUserListsIfNeeded, addList },
)(UserLists)

