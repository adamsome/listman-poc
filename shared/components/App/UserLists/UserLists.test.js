/* eslint-disable import/no-extraneous-dependencies */

import React from 'react'
import { shallow, mount } from 'enzyme'

import { UserLists } from './UserLists'
import UserPage from './UserPage'
import Error404 from '../../Error404'

it('renders correctly', () => {
  const wrapper = shallow(
    <UserLists
      isLoading={false}
      user={{
        username: "test_username",
        description: "description of user",
        avatar_url: "http://path/to/avatar",
      }}
      lists={[
        { title: "list title" }
      ]}
    />
  )
  expect(wrapper.find(UserPage)).toHaveLength(1)
  expect(wrapper).toMatchSnapshot()
})

it('renders <Error404> on error', () => {
  const wrapper = shallow(
    <UserLists error="User not found"/>
  )
  expect(wrapper.find(Error404)).toHaveLength(1)
  expect(wrapper).toMatchSnapshot()
})

it('fetches on mount', () => {
  const mockFetch = jest.fn(() => Promise.resolve())
  const mockFetch2 = jest.fn(() => Promise.resolve())
  const wrapper = mount(
    <UserLists
      fetchUserListsIfNeeded={mockFetch}
      fetchUserIfNeeded={mockFetch2}
      params={{username: "test_username"}}
      user={{
        username: "test_username",
        description: "description of user",
        avatar_url: "http://path/to/avatar",
      }}
      lists={[
        { title: "list title" }
      ]}
    />
  )
  expect(mockFetch.mock.calls.length).toBe(1)
})
