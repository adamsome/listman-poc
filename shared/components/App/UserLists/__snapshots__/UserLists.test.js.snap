exports[`test renders <Error404> on error 1`] = `
<Error404
  error="User not found" />
`;

exports[`test renders correctly 1`] = `
<UserPage
  addList={[Function]}
  isLoading={false}
  lists={
    Array [
      Object {
        "title": "list title",
      },
    ]
  }
  user={
    Object {
      "avatar_url": "http://path/to/avatar",
      "description": "description of user",
      "username": "test_username",
    }
  } />
`;
